torch/csrc/distributed/rpc/profiler/server_process_global_profiler.cpp

```diff
@@ -0,0 +1,96 @@
+#include <shared_mutex>
+
+#include <torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+namespace profiler {
+namespace processglobal {
+
+namespace {
+class StateStackEntry;
+
+#if defined(__MACH__)
+// Compiler error: 'shared_timed_mutex' is unavailable: introduced in
+// macOS 10.12
+using mutexType = std::mutex;
+using rLockType = std::unique_lock<std::mutex>;
+using wLockType = std::unique_lock<std::mutex>;
+#else
+using mutexType = std::shared_timed_mutex;
+using rLockType = std::shared_lock<std::shared_timed_mutex>;
+using wLockType = std::unique_lock<std::shared_timed_mutex>;
```

Comment:
macOS uses two unique_locks but other platforms use a shared_lock + a unique_lock. Is this intentional. If yes, why?

Reply:
If these are read-write locks, the read lock should be shared_lock as well for mac?

Reply:
The macOS versiono PyTorch CI uses is rather old.
`shared_lock` is not avaialble.

It shows this error
```
// Compiler error: 'shared_lock' is unavailable: introduced in
// macOS 10.12
```
We can remove this macro branching only after we abandon old macOS support.

Reply:
Let's keep this as-is. IIUC, there is no correctness problem here, and just a little slower if not using shared_lock for rlock, right?

Reply:
It's a bit slower for macOS. But if a user is using macOS for distributed training, I guess it's not a real use case in my understanding.